### ISSUE TITLE
HGT implementation bug?

### ISSUE THREAD

**Issue by YuxianMeng** [dmlc/dgl#2785]
https://github.com/dmlc/dgl/blob/master/examples/pytorch/hgt/model.py#L81

The implemented hgt set value of nodes multiple times when calculating by different edge attention, which will cause the `nodes["v"]` of earlier etype be overwritten by later one.

**Comment by YuxianMeng** [comment]
@yzh119 Hi, I think [the attention scores](https://github.com/dmlc/dgl/blob/master/examples/pytorch/hgt/model.py#L87)  are also relatoin-specific, which is not fixed in this commit.

**Comment by yzh119** [comment]
Thanks @YuxianMeng , will fix it soon.

**Comment by yzh119** [comment]
I checked the implementation again, though attention scores are relation-specific, we only write the edge data frame on relation-specific `sub_graph` which is not shared among relation types like srcdata/dstdata.

For example:
```python
import dgl
import torch as th

hg = dgl.heterograph(
    {
        ('0', 'e0', '2'): ([0, 0, 1, 1, 2, 2], [0, 1, 2, 0, 1, 2]),
        ('0', 'e1', '1'): ([0, 0, 0, 1, 1, 2], [0, 1, 2, 0, 1, 2]),
        ('1', 'e2', '2'): ([0, 0, 0, 0, 1, 2], [0, 1, 2, 0, 1, 2]),
    }
)

hg['e0'].srcdata['h'] = th.rand(3, 4)
hg['e0'].dstdata['h'] = th.rand(3, 4)
hg['e0'].edata['w'] = th.rand(6, 4)

g0 = hg['e0']
print(g0.srcdata, g0.dstdata, g0.edata)
g1 = hg['e1']
print(g1.srcdata, g1.dstdata, g1.edata)
g2 = hg['e2']
print(g2.srcdata, g2.dstdata, g2.edata)
```

output:
```
{'h': tensor([[0.1746, 0.2506, 0.1820, 0.0042],
        [0.8314, 0.2603, 0.4976, 0.1863],
        [0.9982, 0.1918, 0.0987, 0.5998]])} {'h': tensor([[0.7817, 0.9421, 0.2825, 0.0765],
        [0.5181, 0.2302, 0.0174, 0.4199],
        [0.6569, 0.7463, 0.1506, 0.2798]])} {'w': tensor([[0.9174, 0.5857, 0.7088, 0.6117],
        [0.1625, 0.4203, 0.5710, 0.8732],
        [0.5593, 0.8381, 0.2030, 0.4913],
        [0.1883, 0.7002, 0.7230, 0.2242],
        [0.1233, 0.0457, 0.3056, 0.0831],
        [0.4483, 0.3926, 0.8591, 0.4810]])}
{'h': tensor([[0.1746, 0.2506, 0.1820, 0.0042],
        [0.8314, 0.2603, 0.4976, 0.1863],
        [0.9982, 0.1918, 0.0987, 0.5998]])} {} {}
{} {'h': tensor([[0.7817, 0.9421, 0.2825, 0.0765],
        [0.5181, 0.2302, 0.0174, 0.4199],
        [0.6569, 0.7463, 0.1506, 0.2798]])} {}
```

**Comment by fmellomascarenhas** [comment]
@yzh119, I understand that the code purpose might be to implement HGT, but not specifically in the most efficient way. At first glance, it seems to me that that at least K and Q are being calculated multiple times unnecessarily, for example every time the src or dst node is repeated.

It also seems that possibly some K, Q, V features are being stored even after they were used to calculate the attention and message, consuming unnecessary memory.

Am I understanding it correctly? Thanks!

**Comment by yzh119** [comment]
I've replied to you in another thread.

**Comment by YuxianMeng** [comment]
@yzh119 Thanks for your explanation!

**Comment by Maybewuss** [comment]
> https://github.com/dmlc/dgl/blob/master/examples/pytorch/hgt/model.py#L81
> 
> The implemented hgt set value of nodes multiple times when calculating by different edge attention, which will cause the `nodes["v"]` of earlier etype be overwritten by later one.

Does `nodes["v"]` be modified during the loop ? Even `nodes["v"]` is overwritten by later one, the later `nodes["v"]` is same about early one.

**Comment by fmellomascarenhas** [comment]
@Maybewuss , you are right that is the same v calculated,  but later it is transformed by the edge type, and then stored.

**Comment by Maybewuss** [comment]
> @Maybewuss , you are right that is the same v calculated, but later it is transformed by the edge type, and then stored.

You are right, thanks!

**Comment by Maybewuss** [comment]
> @Maybewuss , you are right that is the same v calculated, but later it is transformed by the edge type, and then stored. This is why the change was made to store v_edge id.

Does `nodes["k"]` have the same question ?

**Comment by Sherler** [comment]
@yzh119 Accroding to the original paper and author's implemtation, though attention scores are relation-specified, normalization should be considered over all of the neighbours. which seems to be different from the current implementation.

[https://github.com/acbull/HGT-DGL/blob/master/model.py#L73](https://github.com/acbull/HGT-DGL/blob/master/model.py#L73)
![image](https://user-images.githubusercontent.com/11971198/122924996-fb53e080-d398-11eb-9455-da3a1bcb6557.png)
![image](https://user-images.githubusercontent.com/11971198/122925070-0f97dd80-d399-11eb-88dc-4b48a6af9a89.png)
